### PR TITLE
feat(doppler): first-class Doppler support for Swarm external secrets

### LINKED ISSUE #171: [RFE] Support for Doppler
Doppler support for swarm external secrets plugin and smoke tests along with docs

### PR DESCRIPTION
**Pull request by atharvamhaske**
## Type of change

- [x] Refactor
- [x] New feature
- [x] CI
- [ ] Bug fix
- [ ] Optimization
- [ ] Documentation

## Mention the secrets provider

**Doppler**

## Description

Adds **Doppler** as a first-class secrets provider for Docker Swarm using the Doppler REST API (`/v3/configs/config/secrets/download`).

### Highlights

- Added `DopplerProvider` implementation with support for both **service tokens** and **CLI tokens**.
- Added configurable config download caching via `DOPPLER_CACHE_TTL` (default: `30s`).
- Rotation polling bypasses the cache to ensure updated secrets are detected immediately.
- Added secret name fallback that automatically uppercases the Docker secret name when the `doppler_secret_name` label is not provided.
- Aligned the provider implementation with the refactor introduced in **PR #147**:
  - `SecretInfo` is constructed once at the driver boundary.
  - Provider-specific implementations now expose `BuildSecretPath()` and `GetSecretFieldLabel()`.
- Added a lightweight mock Doppler API server for local development and CI.
- Added end-to-end smoke tests that verify:
  - secret retrieval
  - automatic rotation
  - Docker Swarm secret updates
- Updated plugin configuration, environment variables, and documentation for the new provider.

### Why we need mock Doppler server in local dev ??

Doppler is SaaS-only, unlike Vault, OpenBao, or AWS (LocalStack), there’s no official
local Docker image or emulator. The Go mock (scripts/tests/mock-doppler-server) stubs the single API endpoint the
plugin uses so smoke tests and CI can run end-to-end without a Doppler account or service token.

### Smoke test (mock Doppler API)

```bash
sudo mkdir -p /run/swarm-external-secrets
sudo chmod 777 /run/swarm-external-secrets

docker swarm init    # if needed

bash scripts/tests/smoke-test-doppler.sh
```

### Plugin configuration (mock server)

```bash
docker plugin set swarm-external-secrets:latest \
    SECRETS_PROVIDER="doppler" \
    DOPPLER_TOKEN="dp.st.smoke-test" \
    DOPPLER_API_URL="http://127.0.0.1:18080" \
    DOPPLER_CACHE_TTL="5s" \
    ENABLE_ROTATION="true" \
    ROTATION_INTERVAL="10s"
```

### Stack configuration

```yaml
secrets:
  smoke_secret:
    driver: swarm-external-secrets:latest
    labels:
      doppler_secret_name: "SMOKE_TEST_PASSWORD"
```

## Screenshots & Logs

N/A.

The smoke test reports pass/fail automatically. Secret rotation can be verified using:

```bash
docker service logs smoke-doppler_app
```

or

```bash
docker exec <container-id> cat /run/secrets/smoke_secret
```

before and after updating the mock server secret.

## Related Tickets & Documents

- Related Issue #171 
- Closes #171 

## Was this PR authored or co-authored using generative AI tooling?

**Yes** — portions of the implementation and documentation were assisted using generative AI tooling.

## Summary by CodeRabbit

* **New Features**
  * Added Doppler as a supported secrets provider with support for cached downloads and rotation.
  * Added Doppler webhook–based rotation triggers, including configurable webhook port/path and signature verification.
  * Added setup examples for Docker Swarm usage, including Doppler secret-name mapping behavior.
* **Documentation**
  * Updated multi-provider documentation and the main README to include Doppler provider details and environment variables.
* **Tests**
  * Added end-to-end smoke tests and automated coverage for Doppler authentication, caching/refresh behavior, fallback naming, and rotation (including webhook flow).

### CODE REVIEW

**Comment by atharvamhaske** [comment]
@coderabbit please review this

File: .github/workflows/smoke-tests.yml [hunk 1]
```diff
@@ -65,4 +65,19 @@ jobs:
...
+
+      - name: Set up Docker
+        uses: docker/setup-buildx-action@v3
+
+      - name: Init Docker Swarm
+        run: docker swarm init
+
+      - name: Run smoke test for doppler
+        run: bash scripts/tests/smoke-test-doppler.sh
+
```
Comment (sanjay7178) [inline .github/workflows/smoke-tests.yml]: no need this , you can move doppler smoke test to `smoke-test` job via matrix strategy

File: scripts/tests/mock-doppler-server/main.go
Comment (sanjay7178) [inline scripts/tests/mock-doppler-server/main.go:1]: I can setup doppler account instead and can use that in CI ,why we need mock server ?

**Review by sanjay7178 — CHANGES_REQUESTED** [review]

**Comment by sanjay7178** [comment]
Other than the `providers/doppler.go` file , other file changes are looks good to me .

File: scripts/tests/mock-doppler-server/main.go
Comment (atharvamhaske) [reply scripts/tests/mock-doppler-server/main.go:1]: i have already put it out in pr, adding here as well:

### Why we need mock Doppler server in local dev ??
Doppler is SaaS-only, unlike Vault, OpenBao, or AWS (LocalStack), there’s no official
local Docker image or emulator. The Go mock (scripts/tests/mock-doppler-server) stubs the single API endpoint the
plugin uses so smoke tests and CI can run end-to-end without a Doppler account or service token.

cc: @sanjay7178

File: .github/workflows/smoke-tests.yml (hunk 1 above)
Comment (atharvamhaske) [reply .github/workflows/smoke-tests.yml]: sure

File: scripts/tests/mock-doppler-server/main.go
Comment (sanjay7178) [reply scripts/tests/mock-doppler-server/main.go:1]: I mean we can use a doppler's SaaS free tier offering to use in CI via doppler CLI through smoke-tests  , Is it sounds good ?

**Comment by sanjay7178** [comment]
Also doppler webhook integration is missing , I would try to merge before this PR #169  .

**Comment by atharvamhaske** [comment]
@coderabbit please review this!

**Review by sanjay7178 — COMMENTED** [review]
## Code Review Summary

### 1. No unit tests for `dopplerwebhook` package
**Severity: Medium**

The `dopplerwebhook` package has zero unit tests. Signature verification, payload parsing, type filtering, and handler routing are only exercised through bash smoke tests. Unit tests would catch regressions faster and cover edge cases the smoke tests don't reach (empty body, missing `X-Doppler-Signature` header, oversized body truncation, non-POST methods, omitted `type` field, etc.).

---

### 2. Unbounded response body read in `downloadSecrets`
**Severity: Low-Medium** | `providers/doppler.go:291`

```go
body, err := io.ReadAll(resp.Body)
```

The Doppler API response is read without a size limit. A misbehaving or compromised server could cause unbounded memory allocation. The webhook handler already uses `io.LimitReader(r.Body, maxBodyBytes)` — the same pattern should be applied here:

```go
body, err := io.ReadAll(io.LimitReader(resp.Body, 10<<20)) // 10 MiB
```

---

### 3. Webhook server starts even when it will have no effect
**Severity: Low** | `driver.go:152-163`

When `DOPPLER_WEBHOOK_ENABLE=true` but rotation is disabled or the provider doesn't support rotation, `startWebhookServer()` logs a warning but still binds the port and accepts connections. The handler then silently discards every event. This could confuse operators who see the port open and assume it's functional. Consider returning early and not starting the server at all:

```go
if config.WebhookEnable && config.ProviderType == "doppler" {
    driver.startWebhookServer()
}
```

---

### 4. `InvalidateCache` defer placement is fragile
**Severity: Low** | `providers/doppler.go:145-149`

```go
func (d *DopplerProvider) InvalidateCache() {
    d.cacheMu.Lock()
    d.cache = make(map[dopplerCacheKey]dopplerCacheEntry)
    defer d.cacheMu.Unlock()
}
```

The `defer` is placed after the map assignment rather than immediately after `Lock()`. This works today, but if anyone adds code between the assignment and the `defer`, it would execute without the unlock being deferred from the lock point. Standard Go convention is `Lock(); defer Unlock()` as adjacent lines.

---

### 5. `CacheInvalidator` comment says "Only for Doppler" but lives in shared interface
**Severity: Info** | `providers/interface.go:39`

```go
// CacheInvalidator(Only for Doppler) is an optional interface...
```

If this is intended as a general optional interface for any caching provider, the "(Only for Doppler)" qualifier should be removed. If it's truly Doppler-specific, it belongs in `providers/doppler.go` rather than `providers/interface.go`.

---

### 6. `dopplerTestState.requestCount` lacks synchronization
**Severity: Low** | `providers/doppler_test.go:247-248`

The `requestCount` field is incremented inside the HTTP handler (which runs in a separate goroutine from the test assertions). The tests happen to pass the race detector because reads and writes don't overlap in practice, but the field should be `atomic.Int32` or protected by the existing mutex to be correct by construction.

**Comment by atharvamhaske** [comment]
@sanjay7178 i have resolved all the changes !
PTAL

**Comment by sanjay7178** [comment]
can you rebase to origin/main instead

**Comment by atharvamhaske** [comment]
SonarCloud failed the quality gate on 6.0% duplication on new code (limit is 3%) — 103 duplicated lines, mostly in tests:

  • `providers/doppler_test.go `— cache + TTL tests repeated the same server/init/GetSecret setup
  • `dopplerwebhook/dopplerwebhook_test.go `— the httptest + POST setup was copied across cases

  Fixed by pulling that into shared helpers (setupDopplerProvider / mustGetSecret / postWebhook). Also cleared the shell smells
  Sonar flagged (local for $1, explicit returns in the Doppler helper functions).

  Behavior is unchanged; just less copy-paste so the gate should pass and no red ci.

**Comment by sanjay7178** [comment]
@atharvamhaske  please resolve the conflicts

**Comment by sanjay7178** [comment]
@codex review

**Comment by sanjay7178** [comment]
@codex review

**Review by sanjay7178 — CHANGES_REQUESTED** [review]
Thanks for the substantial Doppler integration. The provider, webhook tests, mock server, and smoke coverage are a good foundation, but I found several issues that should be addressed before merge:

1. **Rotation polling does not bypass the cache**

   In `driver.go:347-385`, periodic reconciliation eventually calls the normal `GetSecret` path. `providers/doppler.go:240-257` can therefore return a cached value until `DOPPLER_CACHE_TTL` expires.

   For example, with a 10-second rotation interval and one-hour cache TTL, a changed secret may remain undetected for nearly an hour. This also contradicts the PR description stating that rotation polling bypasses the cache.

   Please invalidate once at the beginning of a rotation cycle, or add a fresh-read/conditional-read method specifically for reconciliation. Invalidating once per cycle still allows concurrent secrets from the same config to share one download.

2. **Webhook invalidation races with in-flight downloads**

   `InvalidateCache` only takes `cacheMu`, while downloads are coordinated separately through `fetchMu`. If a webhook invalidates the cache while an older request is in flight, that request can subsequently complete and repopulate the cache with its pre-webhook response. The webhook-triggered check then reads that entry as fresh and misses the update for the entire TTL.

   I reproduced this deterministically. `InvalidateCache` should synchronize with `fetchMu`—using the same lock ordering as the fetch path—or use a cache generation so responses started before invalidation cannot be stored afterward.

3. **Unsigned webhook mode is unsafe on the host network**

   When `DOPPLER_WEBHOOK_SECRET` is empty, every request is accepted, and the listener binds to `:<port>` on the host network. Any reachable caller can repeatedly invalidate the cache and initiate reconciliation/API traffic.

   Please require a signing secret whenever the listener is enabled, or require an explicit insecure-development option rather than silently enabling unsigned production operation.

4. **Prevent bearer-token transmission over plaintext HTTP**

   `validateDopplerAPIURL` accepts arbitrary `http://` endpoints and then sends `Authorization: Bearer ...` to them. Permit HTTP only for loopback test endpoints such as `localhost` and `127.0.0.1`; require HTTPS elsewhere. Doppler's API guidance emphasizes protecting these privileged tokens and documents HTTPS endpoints throughout its API examples: https://docs.doppler.com/reference/api

5. **Add Doppler-aware failure handling**

   `downloadSecrets` currently fails immediately on every non-200 response and discards request/rate-limit metadata. Please add bounded retry with jitter for network failures, 429s, and retryable 5xx responses, honoring `Retry-After`. Doppler documents token-level secret-read limits and the `x-ratelimit-*`/`retry-after` headers in its API reference.

   A structured API error carrying the status, request ID, and parsed messages would also be preferable to returning an arbitrary raw response body.

6. **Consider narrowing downloads to the requested secret**

   Every lookup downloads the complete config, even though the API supports filtering with `secrets=<name>`. This reduces unrelated secret exposure and response size, but the cache key must then include the secret name or use a suitable config-level conditional-cache design.

Dynamic secrets, ETag reconciliation, OIDC authentication, and project/config-aware webhook routing can reasonably remain follow-up work.

Validation performed:

- `go test ./...` passed.
- `go test -race ./providers ./dopplerwebhook` passed.
- The PR currently has a merge conflict in `readme.md`, which also needs resolution before merge.

**Review by sanjay7178 — COMMENTED** [review]
Since this implementation talks directly to Doppler rather than using an SDK, please model the REST API version explicitly in the client. Keep the base URL separate from a code-owned `v3` API prefix and construct endpoints centrally with `url.JoinPath`. The version should not be an environment option because changing it may require different request and response models. This will make a future `v4` migration deliberate and testable rather than requiring endpoint strings to be located and replaced throughout the provider.
